AlgebraInLean/Chapter02/Sheet04.lean

```diff
@@ -4,196 +4,209 @@ set_option linter.unusedTactic false
+
+@[ext] -- allows the `ext` tactic to use this theorem automatically
+theorem Group.Automorphisms.ext {a b : Automorphisms G} : a.val = b.val → a = b := Subtype.ext
+
+noncomputable instance : Group (Group.Automorphisms G) where
+  op a b := {
+    val := a.val ∘ b.val
+    property := by
+      -- SAMPLE SOLUTION
+      exact Group.isomorphism_comp b.prop a.prop
+      -- END SAMPLE SOLUTION
+      done
+  }
+
+  op_assoc a b c := by
+    -- SAMPLE SOLUTION
+    rfl
+    -- END SAMPLE SOLUTION
+    done
+
+  id := {
+    val :=
+      -- SAMPLE SOLUTION
+      id
+      -- END SAMPLE SOLUTION
+    property := by
+      -- SAMPLE SOLUTION
+      exact Group.id_isomorphism
+      -- END SAMPLE SOLUTION
+      done
+  }
+
+  op_id a := by
+    -- SAMPLE SOLUTION
+    rfl
+    -- END SAMPLE SOLUTION
+    done
 
-/- Show that ψ is a group automorphism -/
-theorem ψ_automorphism (g : G) : ∀ x y : G, Conjugate g (μ x y) = μ (Conjugate g x)
-(Conjugate g y) ∧ Bijective (Conjugate g) := by
-  -- SAMPLE SOLUTION
-  intros x y
-  constructor
-  -- Prove homomorphism
-  · unfold Conjugate
-    simp only [op_assoc]
-    rw[← op_assoc (ι g), inv_op, id_op]
-  -- Prove bijectivity
-  · rw[Bijective]
-    constructor
-    -- Injectivity
-    · intros x y h
-      unfold Conjugate at h
-      have h1 : μ (μ g x) (ι g) = μ (μ g y) (ι g) → μ g x = μ g y := by
-        intro h
-        apply right_cancel at h
-        exact h
-      apply h1 at h
-      apply left_cancel at h
-      exact h
-    -- Surjectivity
-    · intro z
-      use μ (ι g) (μ z g)
-      unfold Conjugate
-      simp only [op_assoc]
-      rw [← op_assoc g, op_inv, id_op, op_id]
+  id_op a := by
+    -- SAMPLE SOLUTION
+    rfl
     -- END SAMPLE SOLUTION
-  done
+    done
+
+  inv a := {
+    val := inv_of_bijective a.prop.right
+    property := by
+      -- SAMPLE SOLUTION
+      exact Group.isomorphism_inv a.prop
+      -- END SAMPLE SOLUTION
+      done
+  }
+
+  inv_op a := by
+    -- SAMPLE SOLUTION
+    unfold μ
+    ext : 1
+    simp
+    exact (inv_of_bijective_spec a.prop.right).left
+    -- END SAMPLE SOLUTION
+    done
 
 /-
-
-That proof was tough! But, it was a great exercise for you to prove in the most arbitrary sense,
-since such a proof will be useful later when learning about group actions and automorphism
-groups (meaning this sheet will likely be referenced in later chapters as one to come back to)!
-
+We could also consider automorphisms of other algebraic structures (or even other categories). In
+the most basic case where there are no operations or properties to preserve, we that the
+automorphisms are just bijections from a set to itself, which (in the finite case) is just the
+symmetric group! This means that group-automorphism group is contained within a symmetric group,
+which means that it is a subgroup. The next chapter discusses this relation between groups more
+in-depth.
 -/
-
-
```

Comment:
:(

Reply:
Chapter 1 didn't have a sendoff, so I thought it was weird to have one here. I think that the discussion of subgroups works well. I did keep the "That proof was tough" though, it's just up higher since I added the automorphism group stuff.